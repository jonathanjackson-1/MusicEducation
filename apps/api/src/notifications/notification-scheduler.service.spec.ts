import { NotificationSchedulerService } from './scheduler/notification-scheduler.service';

const jobs: any[] = [];
const prisma: any = {
  notificationJob: {
    create: jest.fn().mockImplementation(async ({ data }: any) => {
      const job = { id: `job-${jobs.length + 1}`, ...data };
      jobs.push(job);
      return job;
    }),
  },
};

describe('NotificationSchedulerService', () => {
  const scheduler = new NotificationSchedulerService(prisma);

  beforeEach(() => {
    jobs.splice(0, jobs.length);
    prisma.notificationJob.create.mockClear();
  });

  it('schedules lesson reminder jobs with offsets', async () => {
    const startAt = new Date('2024-04-22T15:30:00.000Z');
    const result = await scheduler.scheduleLessonReminders({
      studioId: 'studio-1',
      lessonId: 'lesson-1',
      startAt,
      recipients: ['student-1', 'educator-1'],
      reminderOffsetsHours: [24, 1],
      context: { lessonDate: 'April 22', lessonTime: '3:30 PM' },
    });

    expect({
      result: result.map((job: any) => ({
        ...job,
        runAt: job.runAt.toISOString(),
      })),
      persisted: jobs.map((job) => ({
        ...job,
        runAt: job.runAt.toISOString(),
      })),
    }).toMatchSnapshot();
  });

  it('schedules overdue assignment nudge', async () => {
    const dueAt = new Date('2024-04-21T17:00:00.000Z');
    const job = await scheduler.scheduleOverdueAssignmentNudge({
      studioId: 'studio-1',
      assignmentId: 'assignment-1',
      userId: 'student-1',
      dueAt,
      nudgeDays: 2,
      context: { assignmentTitle: 'Chromatic Scales' },
    });

    expect({
      job: { ...job, runAt: job.runAt.toISOString() },
    }).toMatchSnapshot();
  });

  it('schedules weekly practice summary', async () => {
    const weekOf = new Date('2024-04-15T00:00:00.000Z');
    const job = await scheduler.scheduleWeeklyPracticeSummary({
      studioId: 'studio-1',
      userId: 'student-1',
      weekOf,
      context: { totalMinutes: 210 },
    });

    expect({
      job: { ...job, runAt: job.runAt.toISOString() },
    }).toMatchSnapshot();
  });
});
