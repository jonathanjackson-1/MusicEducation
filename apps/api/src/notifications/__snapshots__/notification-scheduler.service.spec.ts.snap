// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationSchedulerService schedules lesson reminder jobs with offsets 1`] = `
{
  "persisted": [
    {
      "channel": null,
      "id": "job-1",
      "payload": {
        "lessonId": "lesson-1",
        "reminderOffsetHours": 24,
        "templateContext": {
          "lessonDate": "April 22",
          "lessonTime": "3:30 PM",
          "reminder": true,
          "reminderWindow": "24 hours",
        },
      },
      "runAt": "2024-04-21T15:30:00.000Z",
      "status": "PENDING",
      "studioId": "studio-1",
      "template": "BOOKING_CONFIRMATION",
      "userId": "student-1",
    },
    {
      "channel": null,
      "id": "job-2",
      "payload": {
        "lessonId": "lesson-1",
        "reminderOffsetHours": 1,
        "templateContext": {
          "lessonDate": "April 22",
          "lessonTime": "3:30 PM",
          "reminder": true,
          "reminderWindow": "1 hour",
        },
      },
      "runAt": "2024-04-22T14:30:00.000Z",
      "status": "PENDING",
      "studioId": "studio-1",
      "template": "BOOKING_CONFIRMATION",
      "userId": "student-1",
    },
    {
      "channel": null,
      "id": "job-3",
      "payload": {
        "lessonId": "lesson-1",
        "reminderOffsetHours": 24,
        "templateContext": {
          "lessonDate": "April 22",
          "lessonTime": "3:30 PM",
          "reminder": true,
          "reminderWindow": "24 hours",
        },
      },
      "runAt": "2024-04-21T15:30:00.000Z",
      "status": "PENDING",
      "studioId": "studio-1",
      "template": "BOOKING_CONFIRMATION",
      "userId": "educator-1",
    },
    {
      "channel": null,
      "id": "job-4",
      "payload": {
        "lessonId": "lesson-1",
        "reminderOffsetHours": 1,
        "templateContext": {
          "lessonDate": "April 22",
          "lessonTime": "3:30 PM",
          "reminder": true,
          "reminderWindow": "1 hour",
        },
      },
      "runAt": "2024-04-22T14:30:00.000Z",
      "status": "PENDING",
      "studioId": "studio-1",
      "template": "BOOKING_CONFIRMATION",
      "userId": "educator-1",
    },
  ],
  "result": [
    {
      "channel": null,
      "id": "job-1",
      "payload": {
        "lessonId": "lesson-1",
        "reminderOffsetHours": 24,
        "templateContext": {
          "lessonDate": "April 22",
          "lessonTime": "3:30 PM",
          "reminder": true,
          "reminderWindow": "24 hours",
        },
      },
      "runAt": "2024-04-21T15:30:00.000Z",
      "status": "PENDING",
      "studioId": "studio-1",
      "template": "BOOKING_CONFIRMATION",
      "userId": "student-1",
    },
    {
      "channel": null,
      "id": "job-2",
      "payload": {
        "lessonId": "lesson-1",
        "reminderOffsetHours": 1,
        "templateContext": {
          "lessonDate": "April 22",
          "lessonTime": "3:30 PM",
          "reminder": true,
          "reminderWindow": "1 hour",
        },
      },
      "runAt": "2024-04-22T14:30:00.000Z",
      "status": "PENDING",
      "studioId": "studio-1",
      "template": "BOOKING_CONFIRMATION",
      "userId": "student-1",
    },
    {
      "channel": null,
      "id": "job-3",
      "payload": {
        "lessonId": "lesson-1",
        "reminderOffsetHours": 24,
        "templateContext": {
          "lessonDate": "April 22",
          "lessonTime": "3:30 PM",
          "reminder": true,
          "reminderWindow": "24 hours",
        },
      },
      "runAt": "2024-04-21T15:30:00.000Z",
      "status": "PENDING",
      "studioId": "studio-1",
      "template": "BOOKING_CONFIRMATION",
      "userId": "educator-1",
    },
    {
      "channel": null,
      "id": "job-4",
      "payload": {
        "lessonId": "lesson-1",
        "reminderOffsetHours": 1,
        "templateContext": {
          "lessonDate": "April 22",
          "lessonTime": "3:30 PM",
          "reminder": true,
          "reminderWindow": "1 hour",
        },
      },
      "runAt": "2024-04-22T14:30:00.000Z",
      "status": "PENDING",
      "studioId": "studio-1",
      "template": "BOOKING_CONFIRMATION",
      "userId": "educator-1",
    },
  ],
}
`;

exports[`NotificationSchedulerService schedules overdue assignment nudge 1`] = `
{
  "job": {
    "channel": null,
    "id": "job-1",
    "payload": {
      "assignmentId": "assignment-1",
      "nudge": true,
      "nudgeDays": 2,
      "templateContext": {
        "assignmentTitle": "Chromatic Scales",
        "nudge": true,
      },
    },
    "runAt": "2024-04-23T17:00:00.000Z",
    "status": "PENDING",
    "studioId": "studio-1",
    "template": "ASSIGNMENT_CREATED",
    "userId": "student-1",
  },
}
`;

exports[`NotificationSchedulerService schedules weekly practice summary 1`] = `
{
  "job": {
    "channel": null,
    "id": "job-1",
    "payload": {
      "templateContext": {
        "totalMinutes": 210,
        "weeklySummary": true,
      },
      "weekOf": 2024-04-15T00:00:00.000Z,
    },
    "runAt": "2024-04-22T18:00:00.000Z",
    "status": "PENDING",
    "studioId": "studio-1",
    "template": "PRACTICE_STREAK",
    "userId": "student-1",
  },
}
`;
